mesonbuild/backend/ninjabackend.py

```diff
@@ -1924,7 +1915,8 @@ def generate_rust_target(self, target: build.BuildTarget) -> None:
 
         args += self.generate_basic_compiler_args(target, rustc, False)
         # Rustc replaces - with _. spaces or dots are not allowed, so we replace them with underscores
-        args += ['--crate-name', target.name.replace('-', '_').replace(' ', '_').replace('.', '_')]
+        crate_name = self._get_rust_dependency_name(target, target.name).replace(' ', '_').replace('.', '_')
+        args += ['--crate-name', crate_name]
```

Comment:
This is wrong. You don't want to pass this crate's name through the `rust_dependency_map`. That will even break things if you want to e.g. link a crate with the same name into this crate (then you'd have to rename the dependency via the map but not this crate itself).

Reply:
The reason I did that is because in my unit test I build the `stuff` crate with different library names: https://github.com/mesonbuild/meson/pull/11714/files#diff-fd68d1be0fd651fd3add1a7ca498172f28c5111ab59ed4b45be1bb8d3a365491R32.

It feels like it should be allowed to have a crate name not matching the lib name, after all that's pretty much the same thing for dependencies. Cargo does not allow that?

Reply:
It allows lib/crate name to be different from the package name (see e.g. `gst-plugin-rtp` as package name vs. `gstrsrtp` as lib/crate name), but in meson terminology that would map to `library` vs. `declare_dependency`.

What you're doing here will disallow certain patterns used by existing crates, most importantly being able to link an older version of the same crate (and giving it a different name locally). I'm using that for example [here](https://github.com/sdroege/ebur128/blob/1b1541c615fe750f7ac56f3f6cfc27b2c2adb53c/Cargo.toml#L26).

You might want a different feature for what you're doing in the test there :)

Reply:
Also quite simply, it makes no sense in cargo to locally change this crate's crate name. It's not observable (apart from logging / backtraces).

For your test you probably want to simply change the `library` name directly instead of such things?

Reply:
I see... so the alternative is to have a separate `rust_crate_name: 'foobar'`, but is there a use-case for that other than my unit test? If not, I guess the alternative is to generate `stuff.rs` in my unit test with a different `#![crate_name = "stuff"]` line for each test, right?

Reply:
I'm not sure why this is even needed. Couldn't you simply pass a different name to `library()` / `build_target()` each time? That is 1:1 mapped to the crate name.

Maybe I'm missing something here :)

Reply:
Removed this commit and adapted the unit test for now.

Reply:
Why is it not sufficient to change the target name though?

Apart from that this looks all good to me.

Reply:
The crate name in stuff.rs is always "stuff" and I cannot build multiple libraries that all have the same name "stuff". Meson complains that multiple targets have the same name because the test matrix build for example a shared library with `rust_crate_type: 'dylib'` and another identical but with `rust_crate_type: 'lib'`.

Reply:
Can't you do `library('stuff_{crate_type}', 'stuff.rs', rust_crate_type: crate_type, ...)`?

Reply:
That fails at build time:
```
FAILED: libstuff-clib-both_libraries.a 
rustc -C linker=cc --color=always -C debug-assertions=yes --crate-type staticlib -g -C relocation-model=pic --crate-name stuff_clib_both_libraries --emit dep-info=stuff-clib-both_libraries.d --emit link -o libstuff-clib-both_libraries.a ../stuff.rs
error: `--crate-name` and `#[crate_name]` are required to match, but `stuff_clib_both_libraries` != `stuff`
 --> ../stuff.rs:1:1
  |
1 | #![crate_name = "stuff"]
  | ^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

Reply:
That means that `library('stuff', ...)` and `#![crate_name = "stuff"]` must match exactly. We can't have a lib name different to crate name.

Reply:
Yeah that makes sense. But why do you have `#![crate_name = "stuff"]` in the code? Without it would work.

Reply:
I was already there, I'm just extending an existing unit test. I did not know that's optional.

Reply:
Ah I see. I'm not exactly sure why it even exists. The answer is probably somewhere in [here](https://doc.rust-lang.org/reference/linkage.html). But for the test, just remove it :)

Also interesting from that link:
> In one session of compilation, the compiler can generate multiple artifacts through the usage of either command line flags [...]

You could build (c)dylib and staticlib/rlib at the same time with the same compiler invocation if you want. Maybe something for a future PR though.

Reply:
So choice would be lib/clib/both? Yes, let's ignore that for now.

Reply:
I updated unit test to remove `#![crate_name = "stuff"]` and it works fine indeed. Thanks.